ampd/src/stellar/verifier.rs

```diff
@@ -15,67 +15,61 @@ impl PartialEq<ContractEventBody> for Message {
         }
 
-        let [symbol, source_address, payload_hash] = &body.topics[..] else {
+        let [symbol, source_address, destination_chain, destination_address, payload_hash] =
+            &body.topics[..]
+        else {
             return false;
         };
 
         let expected_topic: ScVal =
             ScSymbol(StringM::from_str(TOPIC_CALLED).expect("must convert str to ScSymbol")).into();
 
-        let (dest_chain, dest_address) = match &body.data {
-            ScVal::Vec(Some(data)) if data.len() == 3 => {
-                let [dest_chain, dest_address, _] = &data[..] else {
-                    return false;
-                };
-                (dest_chain, dest_address)
-            }
-            _ => return false,
-        };
-
         expected_topic == *symbol
             && (ScVal::Address(self.source_address.clone()) == *source_address)
             && (ScVal::Bytes(self.payload_hash.clone()) == *payload_hash)
-            && (ScVal::String(self.destination_chain.clone()) == *dest_chain)
-            && (ScVal::String(self.destination_address.clone()) == *dest_address)
+            && (ScVal::String(self.destination_chain.clone()) == *destination_chain)
+            && (ScVal::String(self.destination_address.clone()) == *destination_address)
     }
 }
 
 impl PartialEq<ContractEventBody> for VerifierSetConfirmation {
     fn eq(&self, event: &ContractEventBody) -> bool {
         let ContractEventBody::V0(body) = event;
 
-        if body.topics.len() != 1 {
+        if body.topics.len() != 3 {
             return false;
         }
 
-        let [symbol] = &body.topics[..] else {
+        let [symbol, _, signers_hash] = &body.topics[..] else {
             return false;
         };
 
         let expected_topic: ScVal =
             ScSymbol(StringM::from_str(TOPIC_ROTATED).expect("must convert str to ScSymbol"))
                 .into();
 
-        let rotated_signers = match &body.data {
-            ScVal::Vec(Some(data)) if data.len() == 1 => {
-                let [rotated_signers] = &data[..] else {
-                    return false;
-                };
-                rotated_signers.clone()
-            }
-            _ => return false,
+        let weighted_signers = match WeightedSigners::try_from(&self.verifier_set) {
+            Ok(ws) => ws,
+            Err(_) => return false,
+        };
+
+        let hash_bytes = match weighted_signers.hash() {
+            Ok(bytes) => bytes,
+            Err(_) => return false,
+        };
+
+        let bytes = match BytesM::try_from(hash_bytes) {
+            Ok(b) => b,
+            Err(_) => return false,
         };
 
-        WeightedSigners::try_from(&self.verifier_set)
-            .ok()
-            .and_then(|signers| ScVal::try_from(signers).ok())
-            .map_or(false, |signers: ScVal| {
-                symbol == &expected_topic && signers == rotated_signers
-            })
+        let weighted_signers_hash = ScVal::Bytes(ScBytes(bytes));
```

Comment:
```suggestion
        let Some(weighted_signers_hash) = WeightedSigners::try_from(&self.verifier_set)
            .ok()
            .and_then(|weighted_signers| weighted_signers.hash().ok())
            .and_then(|signers_hash| BytesM::try_from(signers_hash).ok())
            .map(ScBytes)
            .map(ScVal::Bytes) else {
                return false
            };
```
a better refactor for later would be to create a PartialEqResult trait type whose `eq` returns a Result type, so we can use `?`, and then `PartialEq` can be implemented by calling this and returning false on error

Reply:
Thank you for the comment. it looks much better, updated.